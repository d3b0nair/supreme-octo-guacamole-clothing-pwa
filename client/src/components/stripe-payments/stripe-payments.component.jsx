import React, { useEffect } from "react";
import CheckoutForm from "../stripe-checkout/stripe-checkout.component";
import { Elements } from "@stripe/react-stripe-js";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import { selectCartTotal } from "../../redux/cart/cart.selector";
import { selectCurrentUser } from "../../redux/user/user.selector";
import { selectClientSecret } from "../../redux/stripe/stripe.selectors";
import { createOrder_START } from "../../redux/checkout/checkout.actions";
import { selectCartItems } from "../../redux/cart/cart.selector";
import {
  createPaymentIntent_start,
  updatePaymentIntent_start,
  cancelPaymentIntent_start,
} from "../../redux/stripe/stripe.actions";
import { selectCheckoutStatus } from "../../redux/checkout/checkout.selectors";
import { selectStripePayimentIntentId } from "../../redux/stripe/stripe.selectors";
import { ElementsContainer, ELEMENTS_OPTIONS } from "./stripe-payments.styles";
import { loadStripe } from "@stripe/stripe-js";

const stripePromise = loadStripe(
  "pk_test_51IQnEYB1K9Yo1Z8kLlGiGgGyaJwhHqvwWFnzi81Bs63AmfXmkMooVeAdrV7YiVwqjwa1Pqck5nzaAj5n6wf8wAFf00EYWjrM5p"
);

export function StripePayments({
  cartItems,
  checkoutStatus,
  createPaymentIntent,
  updatePaymentIntent,
  paymentIntentId,
  total,
  userId,
  clientSecret,
  createOrder,
  cancelPaymentIntent,
}) {
  useEffect(() => {
    if (paymentIntentId === null && checkoutStatus === "START") {
      createPaymentIntent({ cartItems });
    } else if (paymentIntentId && checkoutStatus === "START") {
      updatePaymentIntent({
        id: paymentIntentId,
        update: { items: cartItems },
      });
    }
  }, [
    checkoutStatus,
    cartItems,
    createPaymentIntent,
    paymentIntentId,
    updatePaymentIntent,
  ]);

  return (
    <ElementsContainer>
      <Elements stripe={stripePromise} options={ELEMENTS_OPTIONS}>
        <CheckoutForm
          total={total}
          updatePaymentIntent={updatePaymentIntent}
          cartItems={cartItems}
          checkoutStatus={checkoutStatus}
          paymentIntentId={paymentIntentId}
          clientSecret={clientSecret}
          userId={userId}
          createOrder={createOrder}
          cancelPaymentIntent={cancelPaymentIntent}
        />
      </Elements>
    </ElementsContainer>
  );
}

const mapStateToProps = createStructuredSelector({
  paymentIntentId: selectStripePayimentIntentId,
  cartItems: selectCartItems,
  total: selectCartTotal,
  checkoutStatus: selectCheckoutStatus,
  clientSecret: selectClientSecret,
  userId: selectCurrentUser,
});

const mapDispatchToProps = (dispatch) => ({
  createPaymentIntent: (cartItems) =>
    dispatch(createPaymentIntent_start(cartItems)),
  updatePaymentIntent: (payload) =>
    dispatch(updatePaymentIntent_start(payload)),
  createOrder: (id, items, customer, total, userId) =>
    dispatch(createOrder_START(id, items, customer, total, userId)),
  cancelPaymentIntent: (payload) =>
    dispatch(cancelPaymentIntent_start(payload)),
});

export default connect(mapStateToProps, mapDispatchToProps)(StripePayments);
