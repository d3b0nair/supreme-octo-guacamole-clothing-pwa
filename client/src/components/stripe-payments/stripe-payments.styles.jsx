import styled from "styled-components";
export const ELEMENTS_OPTIONS = {
  fonts: [
    {
      cssSrc: "https://fonts.googleapis.com/css?family=Roboto",
    },
  ],
};

export const ElementsContainer = styled.div`
  min-width: 350px;
  position: relative;
  @media screen and (max-width: 800px) {
    min-width: auto;
    position: relative;
  }
`;