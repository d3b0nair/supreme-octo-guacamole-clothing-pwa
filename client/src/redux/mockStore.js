import { createStore } from "redux";
import { persistReducer } from "redux-persist";
import { rootReducer, persistConfig } from "./root-reducer";
export const createMockStore = ({ mockState, mockReducers }) => {
  const combinedReducers = persistReducer(
    persistConfig,
    rootReducer(mockReducers)
  );
  return createStore(combinedReducers, mockState);
};
