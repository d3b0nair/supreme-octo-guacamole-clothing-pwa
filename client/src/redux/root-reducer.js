import { combineReducers } from "redux";
import { persistReducer } from "redux-persist";
import storage from "redux-persist/lib/storage";

import userReducer from "./user/user.reducer";
import cartReducer from "./cart/cart.reducer";
import directoryReducer from "./directory/directory.reducer";
import shopReducer from "./shop/shop.reducer";
import checkoutReducer from "./checkout/checkout.reducer";
import stripeReducer from "./stripe/stripe.reducer";
import orderReducer from "./order/order.reducer"

const persistConfig = {
  key: "root",
  storage,
  whitelist: ["cart", "stripe"],
};
const rootReducer = combineReducers({
  user: userReducer,
  cart: cartReducer,
  checkout: checkoutReducer,
  directory: directoryReducer,
  shop: shopReducer,
  stripe: stripeReducer,
  selectedOrder: orderReducer
});
export default persistReducer(persistConfig, rootReducer);
