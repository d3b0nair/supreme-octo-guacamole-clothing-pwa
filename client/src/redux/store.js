import { createStore, applyMiddleware, compose } from "redux";
import { persistStore } from "redux-persist";
import logger from "redux-logger";
import createSagaMiddleware from "redux-saga";

import rootReducer from "./root-reducer";
import rootSaga from "./root-saga";

const sagaMiddleware = createSagaMiddleware();

const middlewares = [sagaMiddleware];
const composeEnhancers =
  (process.env.NODE_ENV !== "production" &&
    typeof window !== "undefined" &&
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) ||
  compose;
if (process.env.NODE_ENV === "development") {
  middlewares.push(logger);
}
export const store = createStore(
  rootReducer,
  composeEnhancers(applyMiddleware(...middlewares))
);

sagaMiddleware.run(rootSaga);

export const persistor = persistStore(store);
