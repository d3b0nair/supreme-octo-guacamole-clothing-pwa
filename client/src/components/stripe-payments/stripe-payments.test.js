import React from "react";
import { render, cleanup } from "@testing-library/react";
import { createMockStore } from "../../redux/mockStore";
import { Provider } from "react-redux";
import { StripePayments } from "./stripe-payments.component";

const mockCheckoutSucceded = jest.fn();
const mockCreatePaymentIntent = jest.fn();
const mockUpdatePaymentIntent = jest.fn();
const mockCreateOrder = jest.fn();
const mockTotal = 1234;
const mockReducer = (state = {}, action) => state;
const store = createMockStore({
  state: {},
  reducers: { testReducer: mockReducer },
});
const mockProps = (otherProps) => ({
  ...otherProps,
  checkoutSucceded: mockCheckoutSucceded,
  createPaymentIntent: mockCreatePaymentIntent,
  updatePaymentIntent: mockUpdatePaymentIntent,
  createOrder: mockCreateOrder,
  cartItems: [
    {
      id: 111,
      name: "Brown Brimm",
      price: 35,
      imageUrl: "https://i.ibb.co/ZYW3VTp/brown-brim.png",
      quantity: 4,
    },
    {
      imageUrl: "https://i.ibb.co/ypkgK0X/blue-beanie.png",
      id: 2,
      price: 18,
      name: "Blue Beanie",
      quantity: 4,
    },
  ],
  checkoutStatus: "START",
  total: mockTotal,
  userId: {},
  clientSecret: "1234",
});
const component = (mockProps) => (
  <Provider store={store}>
    <StripePayments {...mockProps} />
  </Provider>
);

afterEach(cleanup);

describe("StripePayments component", () => {
  it("should mount", () => {
    const { container } = render(component(mockProps()));
    expect(container).toMatchSnapshot();
  });
  describe("when component mounts", () => {
    it("if paymentIntentId is already assigned the component should call updatePaymentIntent", () => {
      render(component(mockProps({ paymentIntentId: 1234 })));
      expect(mockUpdatePaymentIntent).toBeCalled();
    });
    it("if paymentIntentId is null it should call createPaymentIntent", () => {
      render(component(mockProps({ paymentIntentId: null })));
      expect(mockCreatePaymentIntent).toBeCalled();
    });
  });
});
