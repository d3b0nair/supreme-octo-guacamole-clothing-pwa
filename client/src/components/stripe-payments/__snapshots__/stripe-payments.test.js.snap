// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StripePayments component should mount 1`] = `
<div>
  <div
    class="sc-crHmcD jhDroC"
  >
    <form
      class="sc-iCfMLu"
    >
      <fieldset
        class="sc-furwcr fHpYvk"
      >
        <div
          class="sc-bdvvtL hMBGla"
        >
          <input
            autocomplete="name"
            class="sc-gsDKAQ hVBime"
            data-testid="FormInputContainer"
            id="name"
            required=""
            type="text"
          />
          <label
            class="sc-dkPtRN jpMQjL  form-input-label"
            data-testid="FormInputLabel"
          >
            Full Name
          </label>
        </div>
        <div
          class="sc-bdvvtL hMBGla"
        >
          <input
            autocomplete="email"
            class="sc-gsDKAQ hVBime"
            data-testid="FormInputContainer"
            id="email"
            required=""
            type="email"
          />
          <label
            class="sc-dkPtRN jpMQjL  form-input-label"
            data-testid="FormInputLabel"
          >
            Email Address
          </label>
        </div>
        <div
          class="sc-bdvvtL hMBGla"
        >
          <input
            autocomplete="tel"
            class="sc-gsDKAQ hVBime"
            data-testid="FormInputContainer"
            id="phone"
            required=""
            type="tel"
          />
          <label
            class="sc-dkPtRN jpMQjL  form-input-label"
            data-testid="FormInputLabel"
          >
            Phone number
          </label>
        </div>
      </fieldset>
      <fieldset
        class="sc-furwcr fHpYvk"
      >
        <div
          class="sc-bdvvtL hMBGla"
        >
          <input
            autocomplete="street-address"
            class="sc-gsDKAQ hVBime"
            data-testid="FormInputContainer"
            id="street"
            required=""
            type="street"
          />
          <label
            class="sc-dkPtRN jpMQjL  form-input-label"
            data-testid="FormInputLabel"
          >
            Street Address
          </label>
        </div>
        <div
          class="sc-bdvvtL hMBGla"
        >
          <input
            class="sc-gsDKAQ hVBime"
            data-testid="FormInputContainer"
            id="city"
            required=""
            type="city"
          />
          <label
            class="sc-dkPtRN jpMQjL  form-input-label"
            data-testid="FormInputLabel"
          >
            City
          </label>
        </div>
        <div
          class="sc-iqseJM ZTRnn"
        >
          <div
            class="sc-bdvvtL hMBGla"
          >
            <input
              class="sc-gsDKAQ hVBime"
              data-testid="FormInputContainer"
              id="state"
              required=""
              type="state"
            />
            <label
              class="sc-dkPtRN jpMQjL  form-input-label"
              data-testid="FormInputLabel"
            >
              State
            </label>
          </div>
          <div
            class="sc-bdvvtL hMBGla"
          >
            <input
              class="sc-gsDKAQ hVBime"
              data-testid="FormInputContainer"
              id="zip"
              required=""
              type="zip"
            />
            <label
              class="sc-dkPtRN jpMQjL  form-input-label"
              data-testid="FormInputLabel"
            >
              Zip Code
            </label>
          </div>
        </div>
      </fieldset>
      <fieldset
        class="sc-furwcr fHpYvk"
      >
        <div
          class="sc-pVTFL hObpPb"
        >
          <div
            class="sc-kDTinF gpvLKs"
          >
            <div />
          </div>
        </div>
      </fieldset>
      <button
        class="sc-hKwDye hzDzeB sc-jrQzAO dsEfPy"
        type="submit"
      >
        Pay 1234$
      </button>
    </form>
  </div>
</div>
`;
